Add ZBMINI device support and extend ZBMINIR2 clusters

s-dimaio opened the pull request:
This pull request introduces the following changes:

(i) added support for the ZBMINI device.
(ii) extended ZBMINIR2 functionality by adding the following clusters:
- TurboMode
- network_led
- power_on_delay_state
- power_on_delay_time
- switch_mode
- detach_mode
(iii) implemented inching support for ZBMINIR2.

These enhancements improve compatibility and provide more control options for both ZBMINI and ZBMINIR2 devices.

Let me know if any changes or improvements are needed!

Simone

s-dimaio committed:
feat: add ZBMINI device and new clusters for ZBMINIR2

gpmachado commented:
Tested Turbo Mode (9) on a local build of the forked version:
https://github.com/s-dimaio/Homey.Sonoff.Zigbee/tree/add/zbmini
Result: Turbo Mode does not seem to work.
UI: No errors displayed.
Network indicator: Worked as expected.
Logs: Show that Turbo Mode is not active/working.
Note: In the original developer’s test version 1.7.8, Turbo Mode triggers an error in the app UI.
This is just to report the behavior for review.

log 
✓ Pre-processing app...
Added FlowCard `sonoff_rf` for type `triggers`
Added FlowCard `on_receive` for type `triggers`
Added Driver `RF`
Added FlowCard `{{driverId}}:single_click` for type `triggers`
Added FlowCard `{{driverId}}:double_click` for type `triggers`
Added FlowCard `{{driverId}}:long_click` for type `triggers`
Added Driver `SNZB-01`
Added FlowCard `{{driverId}}:single_click` for type `triggers`
Added FlowCard `{{driverId}}:double_click` for type `triggers`
Added FlowCard `{{driverId}}:long_click` for type `triggers`
Added Driver `SNZB-01P`
Added Driver `SNZB-02`
Added Driver `SNZB-02D`
Added Driver `SNZB-02P`
Added Driver `SNZB-03`
Added Driver `SNZB-03P`
Added Driver `SNZB-04`
Added Driver `SNZB-04P`
Added Driver `SNZB-05P`
Added FlowCard `is_bright` for type `conditions`
Added Driver `SNZB-06P`
Added Driver `TRVZB`
Added Driver `ZBCurtain`
Added Driver `ZBMINI`
Added Driver `ZBMINIL`
Added Driver `ZBMINIL2`
Added FlowCard `ZBMINIR2:click` for type `triggers`
Added Driver `ZBMINIR2`
Added Capability `sonoff_illuminance`
Added Signal `sonoff_rf` for frequency `433`
Added Signal `sonoff_rf_tx` for frequency `433`
✓ Validating app...
✓ App validated successfully against level `debug`
✓ Creating Remote Debug Session...
✓ Downloading Docker Image...
 — [latest] Pulling from athombv/homey-app-runner
 — [4f4fb700ef54] Pulling fs layer
 — [cd1f6abd6845] Pulling fs layer
 — [d9ffcda75546] Download complete
 — [4f4fb700ef54] Download complete
 — [cd1f6abd6845] Downloading 5%
 — [cd1f6abd6845] Downloading 12%
 — [cd1f6abd6845] Downloading 20%
 — [cd1f6abd6845] Downloading 27%
 — [cd1f6abd6845] Downloading 34%
 — [cd1f6abd6845] Downloading 41%
 — [cd1f6abd6845] Downloading 48%
 — [cd1f6abd6845] Downloading 55%
 — [cd1f6abd6845] Downloading 62%
 — [cd1f6abd6845] Downloading 69%
 — [cd1f6abd6845] Downloading 75%
 — [cd1f6abd6845] Downloading 82%
 — [cd1f6abd6845] Downloading 89%
 — [cd1f6abd6845] Downloading 94%
 — [cd1f6abd6845] Downloading 100%
 — [cd1f6abd6845] Download complete
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [cd1f6abd6845] Extracting
 — [4f4fb700ef54] Pull complete
 — [cd1f6abd6845] Pull complete
 — Digest: sha256:e0938084d506efef216c7f7801bc336d705450b32e5b816fb2a4edb8940e15cd
 — Status: Downloaded newer image for ghcr.io/athombv/homey-app-runner:latest
✓ Preparing Environment Variables...
✓ Connecting to `Gabriel's Homey Pro`...
✓ Starting `se.styrahem.sonoff.zigbee.test@1.7.9` remotely...
✓ Starting debugger at 0.0.0.0:9229...
 — Open `about://inspect` in Google Chrome and select the remote target.
✓ Starting `se.styrahem.sonoff.zigbee.test@1.7.9` in a Docker container...
 — Press CTRL+C to quit.
─────────────── Logging stdout & stderr ───────────────
ℹ Warning: The /userdata folder is not synced with Homey Pro while developing.
After running the app from the Homey App Store, your /userdata may be out of sync.
Modified: /Users/gabriel/.athom-cli/apps-userdata/se.styrahem.sonoff.zigbee.test/se.styrahem.sonoff.zigbee.test
Modified: /var/folders/gd/ghtx6pw535zcr64j5rtt6xy80000gn/T/apps-tmp/se.styrahem.sonoff.zigbee.test/se.styrahem.sonoff.zigbee.test
Debugger listening on ws://0.0.0.0:9229/d2f84bfd-af08-4d2a-8dd6-dae19779fc4e
For help, see: https://nodejs.org/en/docs/inspector
2026-01-04T13:52:05.976Z [log] [SonoffZigbeeApp] Sonoff Zigbee - StyraHem, initiating...
2026-01-04T13:54:15.945Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] ZigBeeDevice has been initialized { firstInit: true, isSubDevice: false }
2026-01-04T13:54:15.946Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] NodeInit SonoffBase
2026-01-04T13:54:15.946Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] ------------------------------------------
2026-01-04T13:54:15.946Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Node: 4ee11c1a-8b03-4182-b5c7-476c8b2a5b1b
2026-01-04T13:54:15.946Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] - Receive when idle: true
2026-01-04T13:54:15.947Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] - Endpoints: 1
2026-01-04T13:54:15.947Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] -- Clusters:
2026-01-04T13:54:15.947Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] --- basic
2026-01-04T13:54:15.947Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] --- onOff
2026-01-04T13:54:15.947Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] --- SonoffCluster
2026-01-04T13:54:15.947Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] ------------------------------------------
2026-01-04T13:54:15.956Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] get → onoff → read attribute (cluster: onOff, attributeId: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) read attributes [ 0 ]  +0ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [],
  data: onOff.readAttributes { attributes: [ 0 ] },
  cmdId: 0,
  trxSequenceNumber: 1
} +0ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) configure reporting [
  {
    direction: 'reported',
    attributeId: 0,
    attributeDataType: 16,
    minInterval: 0,
    maxInterval: 3600
  }
]  +2ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [],
  data: onOff.configureReporting { reports: [ [Object] ] },
  cmdId: 6,
  trxSequenceNumber: 2
} +0ms
2026-01-04T13:54:15.960Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Ask attribute [ 'powerOnBehavior' ]
  zigbee-clusters:cluster ep: 1, cl: onOff (6) read attributes [ 16387 ]  +1ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [],
  data: onOff.readAttributes { attributes: [ 16387 ] },
  cmdId: 0,
  trxSequenceNumber: 3
} +0ms
2026-01-04T13:54:15.960Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Ask attribute [
  'TurboMode',
  'network_led',
  'power_on_delay_state',
  'power_on_delay_time',
  'switch_mode',
  'detach_mode'
]
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) read attributes [ 18, 1, 20, 21, 22, 23 ]  +0ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLStandardHeader {
  frameControl: [],
  data: SonoffCluster.readAttributes {
    attributes: [ 18, 1, 20, 21, 22, 23 ]
  },
  cmdId: 0,
  trxSequenceNumber: 1
} +0ms
2026-01-04T13:54:15.961Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Sending inching command: {
  enabled: false,
  mode: 'on',
  time_ms: 1,
  time_half_seconds: 2,
  payload_array: [
      1, 23, 7, 128, 1,
      0,  2, 0,   0, 0,
    146
  ],
  payload_hex: '0117078001000200000092',
  payload_breakdown: {
    cmd: '0x01',
    subcmd: '0x17',
    length: 7,
    seqnum: '0x80',
    mode_byte: '0x01',
    channel: 0,
    time_low: '0x02',
    time_high: '0x00',
    checksum: '0x92'
  }
}
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLMfgSpecificHeader {
  frameControl: [ 'clusterSpecific', 'manufacturerSpecific' ],
  data: SonoffCluster.protocolData {
    data: <Buffer 01 17 07 80 01 00 02 00 00 00 92>
  },
  cmdId: 1,
  trxSequenceNumber: 2,
  manufacturerId: 4742
} +1ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame readAttributesStructured.response onOff.readAttributesStructured.response {
  attributes: <Buffer 00 00 00 10 01>
} +2s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) read attributes result { attributes: <Buffer 00 00 00 10 01> } +2ms
2026-01-04T13:54:17.690Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
2026-01-04T13:54:17.691Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] get → onoff → read attribute (cluster: onOff, attributeId: onOff, endpoint: 1) → parsed result true
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame configureReporting.response onOff.configureReporting.response {
  reports: [
    ConfigureReportingResponse {
      status: 'SUCCESS',
      direction: 'reported',
      attributeId: 0
    }
  ]
} +144ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame configureReporting.response onOff.configureReporting.response {
  reports: [
    ConfigureReportingResponse {
      status: 'SUCCESS',
      direction: 'reported',
      attributeId: 0
    }
  ]
} +2s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) unknown command received: ZCLStandardHeader {
  frameControl: Bitmap [ directionToClient ],
  trxSequenceNumber: 2,
  cmdId: 7,
  data: <Buffer 00>
} { dstEndpoint: 1 } +0ms
  zigbee-clusters:endpoint ep: 1, cl: onOff (6), error while handling frame unknown_command_received {
  meta: { dstEndpoint: 1 },
  frame: ZCLStandardHeader {
    frameControl: Bitmap [ directionToClient ],
    trxSequenceNumber: 2,
    cmdId: 7,
    data: <Buffer 00>
  }
} +0ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) configured reporting [
  ConfigureReportingResponse {
    status: 'SUCCESS',
    direction: 'reported',
    attributeId: 0
  }
] +16ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame readAttributesStructured.response onOff.readAttributesStructured.response {
  attributes: <Buffer 03 40 00 30 00>
} +139ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) read attributes result { attributes: <Buffer 03 40 00 30 00> } +19ms
2026-01-04T13:54:19.590Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Got attr [ 'powerOnBehavior' ] { powerOnBehavior: 'off' }
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame readAttributesStructured.response SonoffCluster.readAttributesStructured.response {
  attributes: <Buffer 12 00 00 29 09 00 01 00 00 10 01 14 00 00 10 00 15 00 00 21 3b 00 16 00 00 20 00 17 00 00 10 00>
} +240ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) read attributes result {
  attributes: <Buffer 12 00 00 29 09 00 01 00 00 10 01 14 00 00 10 00 15 00 00 21 3b 00 16 00 00 20 00 17 00 00 10 00>
} +0ms
2026-01-04T13:54:19.831Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Got attr [
  'TurboMode',
  'network_led',
  'power_on_delay_state',
  'power_on_delay_time',
  'switch_mode',
  'detach_mode'
] {
  TurboMode: 9,
  network_led: true,
  power_on_delay_state: false,
  power_on_delay_time: 59,
  switch_mode: 0,
  detach_mode: false
}
2026-01-04T13:54:19.845Z [err] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Error: Invalid Value Type For Setting: TurboMode
    at Remote Process
    at HomeyClient.emit (/node_modules/@athombv/homey-apps-sdk-v3/lib/HomeyClient.js:44:23)
    at Object.emit (/node_modules/@athombv/homey-apps-sdk-v3/manager/drivers.js:78:54)
    at Object.emit (/node_modules/@athombv/homey-apps-sdk-v3/lib/Driver.js:126:50)
    at SonoffZBMINIR2.setSettings (/node_modules/@athombv/homey-apps-sdk-v3/lib/Device.js:427:41)
    at /app/drivers/ZBMINIR2/device.js:210:14
    at /app/drivers/sonoffbase.js:83:6
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame defaultResponse SonoffCluster.defaultResponse { cmdId: 1, status: 'SUCCESS' } +140ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) unknown command received: ZCLMfgSpecificHeader {
  frameControl: Bitmap [ manufacturerSpecific, directionToClient ],
  manufacturerId: 4742,
  trxSequenceNumber: 2,
  cmdId: 11,
  data: <Buffer 01 00>
} { dstEndpoint: 1 } +0ms
  zigbee-clusters:endpoint ep: 1, cl: SonoffCluster (64529), error while handling frame unknown_command_received {
  meta: { dstEndpoint: 1 },
  frame: ZCLMfgSpecificHeader {
    frameControl: Bitmap [ manufacturerSpecific, directionToClient ],
    manufacturerId: 4742,
    trxSequenceNumber: 2,
    cmdId: 11,
    data: <Buffer 01 00>
  }
} +553ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame defaultResponse SonoffCluster.defaultResponse { cmdId: 1, status: 'SUCCESS' } +2s
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) unknown command received: ZCLMfgSpecificHeader {
  frameControl: Bitmap [ manufacturerSpecific, directionToClient ],
  manufacturerId: 4742,
  trxSequenceNumber: 2,
  cmdId: 11,
  data: <Buffer 01 00>
} { dstEndpoint: 1 } +0ms
  zigbee-clusters:endpoint ep: 1, cl: SonoffCluster (64529), error while handling frame unknown_command_received {
  meta: { dstEndpoint: 1 },
  frame: ZCLMfgSpecificHeader {
    frameControl: Bitmap [ manufacturerSpecific, directionToClient ],
    manufacturerId: 4742,
    trxSequenceNumber: 2,
    cmdId: 11,
    data: <Buffer 01 00>
  }
} +2s
2026-01-04T13:54:21.636Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Inching command sent successfully
2026-01-04T13:54:21.637Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Initial inching settings applied
2026-01-04T13:54:29.587Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → false (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 0,
  trxSequenceNumber: 4
} +8s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +88ms
2026-01-04T13:54:30.443Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → true (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 1,
  trxSequenceNumber: 5
} +767ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 1, status: 'SUCCESS' } +92ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 01> } +29ms
2026-01-04T13:54:30.565Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
2026-01-04T13:54:41.548Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Write attribute { TurboMode: true }
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) write attributes { TurboMode: true }  +11s
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLStandardHeader {
  frameControl: [],
  data: SonoffCluster.writeAttributes { attributes: <Buffer 12 00 29 01 00> },
  cmdId: 2,
  trxSequenceNumber: 3
} +0ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame writeAttributesAtomic.response SonoffCluster.writeAttributesAtomic.response {
  attributes: [ AttributeResponse { status: 'SUCCESS', id: 0 } ]
} +95ms
2026-01-04T13:54:51.922Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Write attribute { network_led: false }
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) write attributes { network_led: false }  +10s
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLStandardHeader {
  frameControl: [],
  data: SonoffCluster.writeAttributes { attributes: <Buffer 01 00 10 00> },
  cmdId: 2,
  trxSequenceNumber: 4
} +0ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame writeAttributesAtomic.response SonoffCluster.writeAttributesAtomic.response {
  attributes: [ AttributeResponse { status: 'SUCCESS', id: 0 } ]
} +88ms
2026-01-04T13:55:47.466Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → false (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 0,
  trxSequenceNumber: 6
} +55s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +353ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +25ms
2026-01-04T13:55:47.848Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
2026-01-04T13:55:48.932Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → true (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 1,
  trxSequenceNumber: 7
} +1s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 1, status: 'SUCCESS' } +273ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 01> } +48ms
2026-01-04T13:55:49.255Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
2026-01-04T13:55:50.680Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → false (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 0,
  trxSequenceNumber: 8
} +1s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 01> } +73ms
2026-01-04T13:55:50.754Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +284ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +25ms
2026-01-04T13:55:51.063Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
2026-01-04T13:55:52.167Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → true (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 1,
  trxSequenceNumber: 9
} +1s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 1, status: 'SUCCESS' } +95ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 01> } +29ms
2026-01-04T13:55:52.293Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
2026-01-04T13:55:56.028Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → false (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 0,
  trxSequenceNumber: 10
} +4s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +101ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +20ms
2026-01-04T13:55:56.149Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
2026-01-04T13:55:57.629Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → true (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 1,
  trxSequenceNumber: 11
} +1s
2026-01-04T13:55:59.415Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → false (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 0,
  trxSequenceNumber: 12
} +2s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 1, status: 'SUCCESS' } +47ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 01> } +37ms
2026-01-04T13:55:59.499Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +294ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +31ms
2026-01-04T13:55:59.829Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
2026-01-04T13:56:01.000Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → true (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 1,
  trxSequenceNumber: 13
} +1s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 1, status: 'SUCCESS' } +92ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 01> } +34ms
2026-01-04T13:56:01.128Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +194ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) unknown command received: ZCLStandardHeader {
  frameControl: Bitmap [ directionToClient ],
  trxSequenceNumber: 12,
  cmdId: 11,
  data: <Buffer 00 00>
} { dstEndpoint: 1 } +0ms
  zigbee-clusters:endpoint ep: 1, cl: onOff (6), error while handling frame unknown_command_received {
  meta: { dstEndpoint: 1 },
  frame: ZCLStandardHeader {
    frameControl: Bitmap [ directionToClient ],
    trxSequenceNumber: 12,
    cmdId: 11,
    data: <Buffer 00 00>
  }
} +2m
2026-01-04T13:56:24.676Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → false (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 0,
  trxSequenceNumber: 14
} +23s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +242ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +31ms
2026-01-04T13:56:24.952Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +2s
2026-01-04T13:56:26.546Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
2026-01-04T13:57:03.075Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → true (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 1,
  trxSequenceNumber: 15
} +37s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 1, status: 'SUCCESS' } +100ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 01> } +26ms
2026-01-04T13:57:03.203Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: true
2026-01-04T13:57:04.403Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] set onoff → false (cluster: onOff, endpoint: 1)
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [ 'clusterSpecific' ],
  data: <Buffer >,
  cmdId: 0,
  trxSequenceNumber: 16
} +1s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame defaultResponse onOff.defaultResponse { cmdId: 0, status: 'SUCCESS' } +108ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +13ms
2026-01-04T13:57:04.525Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
2026-01-04T13:57:12.519Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Write attribute { TurboMode: false }
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) write attributes { TurboMode: false }  +8s
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLStandardHeader {
  frameControl: [],
  data: SonoffCluster.writeAttributes { attributes: <Buffer 12 00 29 00 00> },
  cmdId: 2,
  trxSequenceNumber: 5
} +0ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame writeAttributesAtomic.response SonoffCluster.writeAttributesAtomic.response {
  attributes: [ AttributeResponse { status: 'SUCCESS', id: 0 } ]
} +98ms
2026-01-04T13:57:18.918Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Write attribute { TurboMode: true }
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) write attributes { TurboMode: true }  +6s
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLStandardHeader {
  frameControl: [],
  data: SonoffCluster.writeAttributes { attributes: <Buffer 12 00 29 01 00> },
  cmdId: 2,
  trxSequenceNumber: 6
} +0ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame writeAttributesAtomic.response SonoffCluster.writeAttributesAtomic.response {
  attributes: [ AttributeResponse { status: 'SUCCESS', id: 0 } ]
} +95ms
2026-01-04T13:57:26.100Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Write attribute { network_led: true }
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) write attributes { network_led: true }  +7s
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLStandardHeader {
  frameControl: [],
  data: SonoffCluster.writeAttributes { attributes: <Buffer 01 00 10 01> },
  cmdId: 2,
  trxSequenceNumber: 7
} +1ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame writeAttributesAtomic.response SonoffCluster.writeAttributesAtomic.response {
  attributes: [ AttributeResponse { status: 'SUCCESS', id: 0 } ]
} +103ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) write attributes { powerOnBehavior: 'last_state' }  +32s
  zigbee-clusters:cluster ep: 1, cl: onOff (6) send frame ZCLStandardHeader {
  frameControl: [],
  data: onOff.writeAttributes { attributes: <Buffer 03 40 30 ff> },
  cmdId: 2,
  trxSequenceNumber: 17
} +1ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame writeAttributesAtomic.response onOff.writeAttributesAtomic.response {
  attributes: [ AttributeResponse { status: 'SUCCESS', id: 0 } ]
} +86ms
2026-01-04T13:57:57.935Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Write attribute {}
2026-01-04T13:58:12.410Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Write attribute { power_on_delay_state: true, power_on_delay_time: 2 }
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) write attributes { power_on_delay_state: true, power_on_delay_time: 2 }  +14s
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) send frame ZCLStandardHeader {
  frameControl: [],
  data: SonoffCluster.writeAttributes {
    attributes: <Buffer 14 00 10 01 15 00 21 02 00>
  },
  cmdId: 2,
  trxSequenceNumber: 8
} +1ms
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame writeAttributesAtomic.response SonoffCluster.writeAttributesAtomic.response {
  attributes: [ AttributeResponse { status: 'SUCCESS', id: 0 } ]
} +129ms
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +44s
2026-01-04T13:58:56.368Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
  zigbee-clusters:cluster ep: 1, cl: onOff (6) received frame reportAttributes onOff.reportAttributes { attributes: <Buffer 00 00 10 00> } +95ms
2026-01-04T13:58:56.462Z [log] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] handle report (cluster: onOff, capability: onoff), parsed payload: false
  zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) received frame reportAttributes SonoffCluster.reportAttributes { attributes: <Buffer 16 00 20 00> } +10ms
2026-01-04T14:12:52.710Z [log] [ManagerDrivers] [Driver:RF] Sent global trigger
2026-01-04T14:12:52.711Z [log] [ManagerDrivers] [Driver:RF] RF receive 011100101000100100101000 728928
✖ Disconnected from `Gabriel's Homey Pro`

gpmachado commented:
2026-01-04T13:54:19.845Z [err] [ManagerDrivers] [Driver:ZBMINIR2] [Device:d71e92cd-b102-4866-9394-a50c0083660f] Error: Invalid Value Type For Setting: TurboMode

There is a Turbo Mode issue visible in the debug logs for the ZBMINIR2.
Although no error is shown in the Homey UI, the debug log reports that Turbo Mode is not applied correctly.
The log shows TurboMode = 9, which corresponds to radioPower OFF, even after enabling Turbo Mode.
According to Zigbee2MQTT, Turbo Mode requires writing a numeric value to the radioPower attribute:
OFF → 0x09
ON → 0x14
Handling Turbo Mode as a boolean (true/false) results in the device staying at 0x09.
The driver should map the setting to the correct numeric value when writing the attribute.

https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/sonoff.ts

s-dimaio committed:
chore: remove workspace file and add to .gitignore

s-dimaio committed:
fix(ZBMINIR2): convert numeric cluster values to boolean for settings

s-dimaio commented:
The error occurred on the first boot because I tried to save the initial value of turbo_mode (and other clusters) in the settings, leaving them as numeric values ​​instead of converting them to boolean values. Turbo Mode therefore worked correctly, but when the device was first configured, the setting value was incorrect (it was always true even if turbo mode was not active).
I updated the code to fix the problem and now all works fine.

gpmachado commented:
I’ve done some more work and created a separate branch after additional testing and debugging.
You can check it here: https://github.com/gpmachado/Homey.Sonoff.Zigbee/tree/zbminiR2

There are still two non-critical errors remaining, but overall behavior has improved.

gpmachado commented:
I tried to contact the developer: the GitHub version is outdated (1.7.0), while the Store has 1.7.5 (stable) and 1.7.9 (test).
However, the changelog doesn’t mention whether the Turbo Mode issue was fixed in the Store version.
I’d prefer to have the GitHub repository updated accordingly, instead of relying on the experimental build.

gpmachado commented:
Disclaimer:
I did not test the changes in this branch.
All my tests were performed on a different branch, and the observations here are based on comparing logs and behavior with other implementations (Hubitat / Zigbee spec), not on direct testing of this code.

# ZBMINIR2 Unknown Commands Analysis and Solutions

## Issue Overview

During normal operation, the ZBMINIR2 sends two commands that appear as "unknown" in debug logs but don't affect functionality:

### 1. OnOff Cluster Command ID: 2 (appears as cmdId: 4 in some logs)
```
zigbee-clusters:cluster ep: 1, cl: onOff (6) 
unknown command received: ZCLStandardHeader {
  frameControl: Bitmap [ directionToClient, disableDefaultResponse ],
  trxSequenceNumber: 6,
  cmdId: 2,  // Toggle command
  data: <Buffer 00>
}
```

### 2. Sonoff Cluster (FC11) Command ID: 11
```
zigbee-clusters:cluster ep: 1, cl: SonoffCluster (64529) 
unknown command received: ZCLMfgSpecificHeader {
  frameControl: Bitmap [ manufacturerSpecific, directionToClient ],
  manufacturerId: 4742,  // 0x1286 - Sonoff
  trxSequenceNumber: 2,
  cmdId: 11,
  data: <Buffer 01 00>
}
```

## Root Cause Analysis

Based on research across multiple Zigbee implementations, these commands are **expected behavior** and require specific handling:

### Command 1: Toggle Command (cmdId: 2) - **CRITICAL**

**Source: Zigbee2MQTT Issue #8018** [1]

This command is sent when the ZBMINIR2 is in **Detach Relay Mode**. In this mode:
- The external switch does NOT directly control the relay
- Instead, it sends a generic `genOnOff.toggle` command (cmdId: 2) to the coordinator
- **The coordinator MUST respond with a ZCL Default Response**

According to Zigbee2MQTT issue #8018, when detach_relay_mode is enabled, the ZBMINIR2 sends a toggle command and expects a ZCL Default Response back from the controller.

**Why this is critical:**
- Without the Default Response, the device LED blinks indicating "abnormal connection with gateway"
- The switch behavior becomes erratic
- Events may be duplicated or missed

**Zigbee2MQTT Implementation:** [2]
```typescript
// Z2M controller automatically sends:
zh:controller:endpoint: ZCL command genOnOff.defaultRsp({
  "cmdId": 2,        // Original Toggle command
  "statusCode": 0    // Success
})
```

**Hubitat Driver Implementation:** [3] (lines ~1058)
```groovy
if (descMap.command == '02' && descMap.data?.isEmpty()) {
    // ZBMINIR2 expects a ZCL Default Response when reporting Toggle
    logDebug ' => Detach Relay Mode - contact open or closed ...'
    sendZigbeeCommands(["he raw ${device.deviceNetworkId} 1 1 0x0006 {08 01 0B 02 00}"])
    customPush()  // Trigger button event
}
```

**Frame Format for Default Response:**
```
08 XX 0B 02 00
│  │  │  │  └─ Status: 0x00 (Success)
│  │  │  └──── Original Command ID: 0x02 (Toggle)
│  │  └─────── Command ID: 0x0B (Default Response)
│  └────────── Transaction Sequence Number (from received frame)
└───────────── Frame Control: 0x08 (Server to Client, Disable Default Response)
```

### Command 2: Sonoff FC11 Cluster cmdId: 11 - **Acknowledgment Only**

**Source: Zigbee2MQTT Converter** [4]

This command is an **acknowledgment response** from the device, typically sent after:
- Configuring inching mode
- Writing to FC11 cluster attributes (turbo mode, network indicator, etc.)

**No response required** - this is just a confirmation that the device received and processed the configuration command.

**Data format:**
```
<Buffer 01 00>
│      └─ Value/Parameter
└──────── Status (0x01 typically means "accepted")
```

## FC11 Cluster Attributes (Sonoff Ewelink)

**Source: Zigbee2MQTT sonoff.ts** [4] and **ZHA Quirks** [5]

The FC11 (0xFC11) cluster is manufacturer-specific (Manufacturer ID: 0x1286 / 4742) and contains:

| Attribute | ID | Type | Description |
|-----------|-----|------|-------------|
| networkIndicator | 0x0001 | Boolean | Enable/disable network LED |
| turboMode | 0x0012 | Int16 | Zigbee radio power (9=disabled, 20=enabled) |
| delayedPowerOnState | 0x0014 | Boolean | Delayed power-on |
| delayedPowerOnTime | 0x0015 | Uint16 | Delay time in ms |
| externalTriggerMode | 0x0016 | Uint8 | External switch mode |
| detachRelayMode | 0x0017 | Boolean | **Separate switch from relay** |
| detachRelayMode2 | 0x0019 | Bitmap8 | Multi-gang configuration |

### External Trigger Mode Values:
- `0`: **edge** (default) - Detects press/release edges
- `1`: **pulse** - Detects pulse
- `2`: **following(off)** - Follows switch state (normally off)
- `130` (0x82): **following(on)** - Follows switch state (normally on)

## Implementation Recommendations

### 1. Handle Toggle Command with Default Response

```javascript
// Listen for Toggle commands from ZBMINIR2
this.zclNode.endpoints[1].on('frame', async (endpointId, clusterId, frame, meta) => {
  if (clusterId === 6 && frame.header?.commandId === 2) {
    this.log('[ZBMINIR2] Toggle command received (Detach Relay Mode)');
    
    // CRITICAL: Send ZCL Default Response
    const transactionId = frame.header.transactionSequenceNumber || 0;
    const defaultResponseFrame = Buffer.from([
      0x18,           // Frame Control: Client to Server, Disable Default Response
      transactionId,  // Transaction Sequence Number (from received frame)
      0x0B,           // Command ID: Default Response
      0x02,           // Original Command ID: Toggle
      0x00            // Status: Success
    ]);
    
    try {
      await this.zclNode.endpoints[1].sendFrame(
        6,  // OnOff cluster
        defaultResponseFrame,
        { disableResponse: true, manufacturerSpecific: false }
      );
      
      this.log('[ZBMINIR2] ZCL Default Response sent');
      
      // Optional: Trigger button capability or flow
      await this.triggerButtonEvent();
      
    } catch (error) {
      this.error('[ZBMINIR2] Error sending Default Response:', error);
    }
  }
});
```

### 2. Handle FC11 Cluster Acknowledgments

```javascript
// Listen for FC11 manufacturer-specific commands
this.zclNode.endpoints[1].on('frame', (endpointId, clusterId, frame, meta) => {
  if (clusterId === 64529 && frame.header?.commandId === 11) {
    // This is just an ACK - log and ignore
    this.log('[ZBMINIR2] FC11 configuration acknowledgment received');
    
    if (frame.data && frame.data.length >= 2) {
      const status = frame.data.readUInt8(0);
      const value = frame.data.readUInt8(1);
      this.debug(`FC11 ACK - Status: ${status}, Value: ${value}`);
    }
    
    // No response needed
  }
});
```

### 3. Configure FC11 Cluster

```javascript
// Define FC11 cluster in driver
const FC11_CLUSTER = {
  ID: 0xFC11,
  NAME: 'sonoffEwelink',
  MANUFACTURER_ID: 0x1286,  // 4742
  
  ATTRIBUTES: {
    networkIndicator: { id: 0x0001, type: ZCLDataTypes.bool },
    turboMode: { id: 0x0012, type: ZCLDataTypes.int16 },
    delayedPowerOnState: { id: 0x0014, type: ZCLDataTypes.bool },
    delayedPowerOnTime: { id: 0x0015, type: ZCLDataTypes.uint16 },
    externalTriggerMode: { id: 0x0016, type: ZCLDataTypes.uint8 },
    detachRelayMode: { id: 0x0017, type: ZCLDataTypes.bool }
  }
};

// Read configuration on init
async configureZBMINIR2() {
  try {
    const config = await this.zclNode.endpoints[1].clusters.fc11.readAttributes([
      'detachRelayMode',
      'externalTriggerMode',
      'turboMode'
    ]);
    
    this.log('ZBMINIR2 Configuration:', config);
    
    // Configure reporting for external trigger mode changes
    await this.zclNode.endpoints[1].clusters.fc11.configureReporting({
      externalTriggerMode: {
        minInterval: 0,
        maxInterval: 3600,
        minChange: 1
      }
    });
    
  } catch (error) {
    this.error('Error configuring ZBMINIR2:', error);
  }
}
```

## Known Issues to Consider

### Issue: "Opening" Sensor Always Closed
**Source: ZHA Issues #3579, #4364** [6] [7]

The ZBMINIR2 exposes a binary sensor called "Opening" that should change state when the external switch is triggered. However:
- This sensor frequently gets stuck in "closed" state
- Causes erratic behavior with detach relay mode
- Double-tap detection is unreliable

**Recommendation:** Do not expose the "Opening" sensor to users until this is resolved in firmware.

### Issue: Power-On Behavior Firmware Dependency
**Source: ZHA Issue #3579** [6]

The powerOnBehavior attribute only works correctly on firmware version 1.3.0 or later. Earlier firmware versions may ignore this setting.

## Additional Configuration Options

### Unbind Poll Control Cluster (Optional)
**Source: Hubitat Driver** [3] (lines ~1219)

```javascript
// Prevent unnecessary checkin messages
// Zigbee-herdsman responds to checkin which causes slower polling
await this.zclNode.endpoints[1].unbind('pollControl', this.zclNode.endpoints[1]);
```

The Hubitat driver unbinds the Poll Control cluster (0x0020) to prevent the device from sending checkin messages that can interfere with reporting behavior.

## Device Signature

For reference, the complete ZBMINIR2 signature:

```json
{
  "node_descriptor": {
    "logical_type": 1,
    "manufacturer_code": 4742
  },
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x0002",
      "input_clusters": [
        "0x0000", "0x0003", "0x0004", "0x0006",
        "0x0b05", "0xfc11", "0xfc57"
      ],
      "output_clusters": ["0x0003", "0x0006", "0x0019"]
    }
  }
}
```

## References

[1] Zigbee2MQTT Issue #8018 - "Sonoff ZBMINIR2 missing commandToggle"  
https://github.com/Koenkk/zigbee2mqtt/issues/8018

[2] Zigbee2MQTT Controller Log - Default Response Implementation  
https://github.com/Koenkk/zigbee2mqtt/issues/8018#issuecomment-967516849

[3] Hubitat Tuya Zigbee Switch Driver - ZBMINIR2 Implementation  
https://github.com/kkossev/Hubitat/blob/development/Drivers/Tuya%20Zigbee%20Switch/Tuya_Zigbee_Switch_lib_included.groovy

[4] Zigbee2MQTT Converter - Sonoff Devices (ZBMINIR2 definition)  
https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/sonoff.ts

[5] ZHA Device Handlers - ZBMINIR2 Quirk  
https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/sonoff/zbminir2.py

[6] ZHA Issue #3579 - "Incorrect behaviour for Detached Relay mode"  
https://github.com/zigpy/zha-device-handlers/issues/3579

[7] ZHA Issue #4364 - "Sonoff ZBMINIR2 opening sensor always closed"  
https://github.com/zigpy/zha-device-handlers/issues/4364

## Summary

These "unknown commands" are actually **expected behavior**:

1. **cmdId: 2 (Toggle)** - Requires ZCL Default Response when detach relay mode is active
2. **cmdId: 11 (FC11)** - Acknowledgment only, no response needed

The key to proper ZBMINIR2 support is implementing the ZCL Default Response for the Toggle command. Without it, the device will indicate connection errors and behave erratically.

s-dimaio committed:
v1.8.0: feat: Add MINI-ZB2GS series & ZBMINI drivers and improve ZBMINIR2
- Added support for ZBMINI relay
- Added support for MINI-ZB2GS dual-channel smart switch
- Added support for MINI-ZB2GS-L single-channel smart switch
- Improved ZBMINIR2 driver with full Sonoff custom clusters integration
